librad/src/git/trailer.rs

```diff
@@ -0,0 +1,245 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#[derive(Debug, Clone, Eq, PartialEq, thiserror::Error)]
+pub enum Error<'a> {
+    #[error("The trailers paragraph is missing in the given message")]
+    MissingParagraph,
+
+    #[error(
+        "One or more trailers are not in the parseable format <token><separator><value>: '{0}'"
+    )]
+    Unparsable(&'a str),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Trailer<'a> {
+    token: Token<'a>,
+    values: Vec<&'a str>,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Token<'a>(pub &'a str);
+
+/// Parse the trailers of the given message. It looks up the last paragraph
+/// of the message and attempts to parse each of its lines as a [Trailer].
+/// Fails if no trailers paragraph is found or if at least one trailer
+/// fails to be parsed.
+pub fn parse<'a>(message: &'a str, separators: &'a str) -> Result<Vec<Trailer<'a>>, Error<'a>> {
+    let trailers_paragraph = message
+        .rsplit("\n\n")
+        .next()
+        .ok_or(Error::MissingParagraph)?;
+    match parser::trailers(trailers_paragraph, separators) {
+        Ok(("", trailers)) => Ok(trailers),
+        Ok((rest, _)) => Err(Error::Unparsable(rest)),
+        _ => Err(Error::MissingParagraph),
+    }
+}
+
+//TODO(nuno): Support commented trailers
+pub mod parser {
+    use super::{Token, Trailer};
+    use nom::{
+        bytes::complete::take_while1,
+        character::complete::{line_ending, not_line_ending, one_of, space0, space1},
+        multi::many0,
+        sequence::{delimited, preceded, separated_pair, terminated},
+        IResult,
+    };
+
+    /// Parse all the possible trailers.
+    /// It stops when it can no longer parse valid trailers.
+    pub fn trailers<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Vec<Trailer<'a>>> {
+        many0(|s| trailer(s, separators))(s)
+    }
+
+    /// Parse a trailer, which can have an inlined or multilined value.
+    pub fn trailer<'a>(s: &'a str, separators: &'a str) -> IResult<&'a str, Trailer<'a>> {
+        let parser = separated_pair(token, |s| separator(separators, s), values);
+        let (rest, (token, values)) = parser(s)?;
+        Ok((rest, Trailer { token, values }))
+    }
+
+    /// Parse a trailer token.
+    fn token(s: &str) -> IResult<&str, Token> {
+        take_while1(|c: char| c.is_alphanumeric() || c == '-')(s)
+            .map(|(i, token_str)| (i, Token(token_str)))
+    }
+
+    /// Parse the trailer separator, which can be delimited by spaces.
+    fn separator<'a>(separators: &'a str, s: &'a str) -> IResult<&'a str, char> {
+        delimited(space0, one_of(separators), space0)(s)
+    }
+
+    /// Parse the trailer values, which gathers the value after the separator
+    /// (if any) and possible following multilined values, indented by a
+    /// space.
+    fn values(s: &str) -> IResult<&str, Vec<&str>> {
```

Comment:
Can we return a `NonEmpty` instead? We shouldn't be able to have a `Trailer` without any values right?

Reply:
Wrong actually:

```
 ➜ echo -ne "Subject\n\n#Foo:Bar\nLong-trailer:asd\n qwe\n zxc\nX:\n\n\n"| git interpret-trailers --parse          ~/radicle-link  git-trailers ●
Long-trailer: asd qwe zxc
X:

```

Reply:
I realised that below :sweat_smile: I'll mark this as resolved.